xla/hlo/analysis/hlo_replication_analysis.cc

```diff
@@ -100,7 +100,11 @@ HloReplicationAnalysis::DetermineHloInstructionIsReplicated(
             hlo->GetModule()->config().num_partitions();
         absl::flat_hash_set<int64_t> visited_partitions;
         absl::flat_hash_set<int64_t> visited_replicas;
+        std::vector<int64_t> device_set;
+        std::vector<absl::Span<const int64_t>> device_sets;
+        std::vector<std::vector<int64_t>> device_sets_storage;
         for (const auto& group : hlo->replica_groups()) {
+          device_set.clear();
```

Comment:
Since you clear this at the start of the for-loop, any mutations you make are ignored except for the last iteration.

Because the existing logic is already incorrect even without this PR, it's somewhat hard to review this, as I'm undecided whether bugs are just an extension of the original bug and therefore perhaps acceptable to merge, or whether things like this are considered a new bug and therefore should be fixed in this PR. Instead of fixing this in a separate PR after this PR is merged, I think it makes sense to fix it before this PR. Once your fix is merged, you can merge master into this PR and I can re-review.